File: craftinginterpreters/vm.c
#include "vm.h"

#include <stdarg.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

#include "common.h"
#include "compiler.h"
#include "debug.h"
#include "memory.h"
#include "object.h"
#include "value.h"

VM vm;

static Value clockNative(int arg_cnt, Value *args) {
  return NUMBER_VAL((double)clock() / CLOCKS_PER_SEC);
}

static void defineNative(const char *name, NativeFn f) {
  push(OBJ_VAL(copyString(name, (int)strlen(name))));
  push(OBJ_VAL(newNative(f)));
  tableSet(&vm.globals_, AS_STRING(vm.stack_[0]), vm.stack_[1]);
  pop();
  pop();
}

static void resetStack() {
  vm.stack_top_ = vm.stack_;
  vm.frame_count_ = 0;
}

void initVM() {
  resetStack();
  vm.objects_head_ = NULL;
  initTable(&vm.string_set_);
  initTable(&vm.globals_);

  defineNative("clock", clockNative);
}

void freeVM() {
  freeTable(&vm.string_set_);
  freeTable(&vm.globals_);
  freeObjects();
}

Value readConstantLong() {
  // ip already at operand position
  CallFrame *frame = &vm.frames_[vm.frame_count_ - 1];
  Chunk *chunk = &frame->closure_->function_->chunk_;
  int vidx = (chunk->code[frame->ip_ - chunk->code + 2] << 16) |
             (chunk->code[frame->ip_ - chunk->code + 1] << 8) |
             (chunk->code[frame->ip_ - chunk->code + 0]);
  frame->ip_ += 3;

  return chunk->constants.values[vidx];
}
static Value peek(int distance) { return vm.stack_top_[-1 - distance]; }
static void runtimeError(const char *format, ...) {
  va_list args;
  va_start(args, format);
  vfprintf(stderr, format, args);
  va_end(args);
  fputs("\n", stderr);

  for (int i = vm.frame_count_ - 1; i >= 0; i--) {
    CallFrame *frame = &vm.frames_[i];
    ObjFunction *f = frame->closure_->function_;
    size_t instruction = frame->ip_ - f->chunk_.code - 1;
    fprintf(stderr, "[line %d] in ", f->chunk_.lines[instruction]);

    if (f->name_ == NULL) {
      fprintf(stderr, "script\n");
    } else {
      fprintf(stderr, "%s()\n", f->name_->chars_);
    }
  }

  resetStack();
}

static bool isFalsey(Value v) {
  return IS_NIL(v) || (IS_BOOL(v) && !AS_BOOL(v));
}
static void concatenate() {
  ObjString *b = AS_STRING(pop());
  ObjString *a = AS_STRING(pop());
  int len = a->length_ + b->length_;
  char *chars = ALLOCATE(char, len + 1);
  memcpy(chars, a->chars_, a->length_);
  memcpy(chars + a->length_, b->chars_, b->length_);
  chars[len] = '\0';
  ObjString *ret = takeString(chars, len);

  push(OBJ_VAL(ret));
}

static bool call(ObjClosure *closure, int arg_cnt) {
  if (arg_cnt != closure->function_->arity_) {
    runtimeError("Expect %d arguments but got %d.", closure->function_->arity_,
                 arg_cnt);
    return false;
  }
  if (vm.frame_count_ == FRAMES_MAX) {
    runtimeError("Stack overflow.");
    return false;
  }

  CallFrame *frame = &vm.frames_[vm.frame_count_++];
  frame->closure_ = closure;
  frame->ip_ = closure->function_->chunk_.code;
  frame->slots_ = vm.stack_top_ - arg_cnt - 1;

  return true;
}

static bool callValue(Value callee, int arg_cnt) {
  if (IS_OBJ(callee)) {
    switch (OBJ_TYPE(callee)) {
      case OBJ_CLOSURE:
        return call(AS_CLOSURE(callee), arg_cnt);
      case OBJ_NATIVE: {
        NativeFn native = AS_NATIVE(callee);
        Value ret = native(arg_cnt, vm.stack_top_ - arg_cnt);
        vm.stack_top_ -= arg_cnt + 1;
        push(ret);
        return true;
      }

      default:
        // fall as error
        break;
    }
  }
  runtimeError("Can only call functions and clauses.");
  return false;
}

static ObjUpvalue *captureUpvalue(Value *local) {
  ObjUpvalue *created_upvalue = newUpvalue(local);

  return created_upvalue;
}

static InterpretResult run() {
  // the last frame
  CallFrame *frame = &vm.frames_[vm.frame_count_ - 1];

#define READ_BYTE() (*frame->ip_++)
#define READ_CONSTANT() \
  (frame->closure_->function_->chunk_.constants.values[READ_BYTE()])
#define READ_STRING() AS_STRING(READ_CONSTANT())
#define READ_SHORT() \
  (frame->ip_ += 2, (uint16_t)((frame->ip_[-1] << 8) | (frame->ip_[-2])))
#define BINARY_OP(valuetype, op)                      \
  do {                                                \
    if (!IS_NUMBER(peek(0)) && !IS_NUMBER(peek(1))) { \
      runtimeError("Operands mut be numbers.");       \
      return INTERPRET_RUNTIME_ERROR;                 \
    }                                                 \
    double b = AS_NUMBER(pop());                      \
    double a = AS_NUMBER(pop());                      \
    push(valuetype(a op b));                          \
  } while (false)

  while (true) {
    // offset is ip - start
#ifdef DEBUG_TRACE_EXECUTION
    printf("          ");
    for (Value *slot = vm.stack_; slot < vm.stack_top_; slot++) {
      printf("[ ");
      printValue(*slot);
      printf(" ]");
    }
    printf("\n");
    disassembleInstruction(&frame->function_->chunk_,
                           (int)(frame->ip_ - frame->function_->chunk_.code));
#endif
    uint8_t instruction = READ_BYTE();
    switch (instruction) {
      case OP_CONSTANT: {
        Value constant = READ_CONSTANT();
        push(constant);
        /*printValue(constant);*/
        /*printf("\n");*/
        break;
      }
      case OP_CONSTANT_LONG: {
        Value constant = readConstantLong();
        push(constant);
        /*printValue(constant);*/
        /*printf("\n");*/
        break;
      }
      case OP_NIL:
        push(NIL_VAL);
        break;
      case OP_TRUE:
        push(BOOL_VAL(true));
        break;
      case OP_FALSE:
        push(BOOL_VAL(false));
        break;
      case OP_POP:
        pop();
        break;
      case OP_GET_GLOBAL: {
        // constant idx put in code array
        ObjString *key = READ_STRING();
        Value v;
        if (!tableGet(&vm.globals_, key, &v)) {
          runtimeError("Undefined variable '%s'.", key->chars_);
          return INTERPRET_RUNTIME_ERROR;
        }
        push(v);
        break;
      }
      case OP_SET_GLOBAL: {
        ObjString *key = READ_STRING();
        if (tableSet(&vm.globals_, key, peek(0))) {
          tableDelete(&vm.globals_, key);
          runtimeError("Undefined variable '%s'.", key->chars_);
          return INTERPRET_RUNTIME_ERROR;
        }
        break;
      }
      case OP_GET_LOCAL: {
        uint8_t slot = READ_BYTE();
        push(frame->slots_[slot]);
        break;
      }
      case OP_SET_LOCAL: {
        uint8_t slot = READ_BYTE();
        // name are just for index, they don't have name to search
        // just the idx
        /*vm.stack_[slot] = peek(0);*/
        frame->slots_[slot] = peek(0);
        /* Note that it doesn’t pop the value from the stack. Remember,
         * assignment is an expression, and every expression produces a value.
         * The value of an assignment expression is the assigned value itself,
         * so the VM just leaves the value on the stack.*/
        break;
      }

      case OP_DEFINE_GLOBAL: {
        ObjString *name = READ_STRING();
        tableSet(&vm.globals_, name, peek(0));
        // the value for bind to this variable
        pop();
        break;
      }
      case OP_GET_UPVALUE: {
        uint8_t slot = READ_BYTE();
        push(*frame->closure_->upvalues_[slot]->location_);
        break;
      }
      case OP_SET_UPVALUE: {
        uint8_t slot = READ_BYTE();
        *frame->closure_->upvalues_[slot]->location_ = peek(0);
        break;
      }
      case OP_EQUAL: {
        Value b = pop();
        Value a = pop();
        push(BOOL_VAL(valuesEqual(a, b)));
        break;
      }
      case OP_GREATER: {
        BINARY_OP(BOOL_VAL, >);
        break;
      }
      case OP_LESS: {
        BINARY_OP(BOOL_VAL, <);
        break;
      }
      case OP_NEGATE: {
        if (!IS_NUMBER(peek(0))) {
          runtimeError("Operand must be a number");
          return INTERPRET_RUNTIME_ERROR;
        }
        push(NUMBER_VAL(-AS_NUMBER(pop())));
        break;
      }
      case OP_ADD: {
        if (IS_STRING(peek(0)) && IS_STRING(peek(1))) {
          concatenate();
        } else if (IS_NUMBER(peek(0)) && IS_NUMBER(peek(1))) {
          double b = AS_NUMBER(pop());
          double a = AS_NUMBER(pop());
          push(NUMBER_VAL(a + b));
        } else {
          runtimeError("Operands must be two numbers or two strings.");
          return INTERPRET_RUNTIME_ERROR;
        }
        break;
      }
      case OP_SUBSTRACT: {
        BINARY_OP(NUMBER_VAL, -);
        break;
      }
      case OP_MULTIPLY: {
        BINARY_OP(NUMBER_VAL, *);
        break;
      }
      case OP_DIVIDE: {
        BINARY_OP(NUMBER_VAL, /);
        break;
      }
      case OP_NOT:
        push(BOOL_VAL(isFalsey(pop())));
        break;
      case OP_PRINT:
        printValue(pop());
        printf("\n");
        break;
      case OP_JUMP_IF_FALSE: {
        uint16_t offset = READ_SHORT();
        if (isFalsey(peek(0))) {
          /*vm.ip_ += offset;*/
          frame->ip_ += offset;
        }
        break;
      }
      case OP_JUMP: {
        uint16_t offset = READ_SHORT();
        /*vm.ip_ += offset;*/
        frame->ip_ += offset;
        break;
      }
      case OP_LOOP: {
        uint16_t back = READ_SHORT();
        /*vm.ip_ -= back;*/
        frame->ip_ -= back;
        break;
      }

      case OP_CALL: {
        int arg_cnt = READ_BYTE();
        if (!callValue(peek(arg_cnt), arg_cnt)) {
          return INTERPRET_RUNTIME_ERROR;
        }
        // after evaluation
        frame = &vm.frames_[vm.frame_count_ - 1];
        break;
      }
      case OP_CLOSURE: {
        ObjFunction *f = AS_FUNCTION(READ_CONSTANT());
        ObjClosure *closure = newClosure(f);
        push(OBJ_VAL(closure));
        for (int i = 0; i < closure->upvalue_cnt_; i++) {
          uint8_t is_local = READ_BYTE();
          uint8_t index = READ_BYTE();
          if (is_local) {
            // An OP_CLOSURE instruction is emitted at the end of a function
            // declaration. At the moment that we are executing that
            // declaration, the current function is the surrounding one. That
            // means the current function’s closure is stored in the CallFrame
            // at the top of the callstack. So, to grab an upvalue from the
            // enclosing function, we can read it right from the frame local
            // variable, which caches a reference to that CallFrame.
            //
            // only call can affect the frame shrink or expand
            // OP_CLOSURE is to get this function and push to stack
            // but no call yet, so there's no frame incresing
            // so the top frame is the enclosing one
            closure->upvalues_[i] = captureUpvalue(frame->slots_ + index);
          } else {
            closure->upvalues_[i] = frame->closure_->upvalues_[index];
          }
        }
        break;
      }

      case OP_RETURN: {
        // goodbye, return, we have print now
        /*printValue(pop());*/
        /*printf("\n");*/
        Value ret = pop();
        vm.frame_count_--;
        if (vm.frame_count_ == 0) {
          // the script
          pop();
          return INTERPRET_OK;
        }
        vm.stack_top_ = frame->slots_;
        push(ret);
        frame = &vm.frames_[vm.frame_count_ - 1];
        break;
      }
    }
  }

#undef READ_BYTE
#undef READ_CONSTANT
#undef READ_STRING
#undef READ_SHORT
}

InterpretResult interpret(const char *source) {
// legacy code
#if 0
  Chunk c;
  initChunk(&c);
  if (!compile(source, &c)) {
    freeChunk(&c);
    return INTERPRET_COMPILE_ERROR;
  }
  vm.chunk_ = &c;
  vm.ip_ = vm.chunk_->code;
#endif
  ObjFunction *f = compile(source);
  if (f == NULL) {
    return INTERPRET_COMPILE_ERROR;
  }
  push(OBJ_VAL(f));
  ObjClosure *closure = newClosure(f);
  pop();
  push(OBJ_VAL(closure));
  call(closure, 0);

  // a new frame
  /*CallFrame *frame = &vm.frames_[vm.frame_count_++];*/
  /*frame->closure_->function_ = f;*/
  /*frame->ip_ = f->chunk_.code;*/
  /*frame->slots_ = vm.stack_;  // this line mark the base pointer*/

  return run();
}

#if 0
InterpretResult interpret(Chunk *c) {
  vm.chunk_ = c;
  vm.ip_ = vm.chunk_->code;

  return run();
}
#endif

void push(Value v) {
  *vm.stack_top_ = v;
  vm.stack_top_++;
}

Value pop() {
  vm.stack_top_--;

  return *vm.stack_top_;
}

File: craftinginterpreters-rs/src/vm.rs
use crate::{
    chunk::{Instruction, Value},
    closure::{Closure, ClosureID, Closures, ObjUpvalue},
    compiler::Parser,
    error::LoxError,
    function::{FunctionID, Functions, NativeFn},
    strings::{LoxString, Strings},
};
use cpu_time::ProcessTime;
use lazy_static::lazy_static;
use std::cell::RefCell;
use std::{collections::HashMap, rc::Rc};

struct CallFrame {
    function: FunctionID,
    closure: ClosureID,
    ip: usize,
    slot: usize, // what's this? base?
}

impl CallFrame {
    fn new(function: FunctionID, closure: ClosureID) -> Self {
        CallFrame {
            function,
            closure,
            ip: 0,
            slot: 0,
        }
    }
}

const MAX_FRAMES: usize = 64;
const STACK_SIZE: usize = MAX_FRAMES * (std::u8::MAX as usize) + 1;

pub struct ExecutionState {
    frames: Vec<CallFrame>,
    stack: Vec<Value>,
    globals: HashMap<LoxString, Value>,
    closures: Closures,
    open_upvalues: Vec<Rc<RefCell<ObjUpvalue>>>,
}

lazy_static! {
    static ref BEGIN_OF_PROGRAM: ProcessTime = ProcessTime::now();
}

fn clock(_args: &[Value]) -> Value {
    Value::Number(BEGIN_OF_PROGRAM.elapsed().as_secs_f64())
}

impl ExecutionState {
    pub fn new(strings: &mut Strings) -> Self {
        let mut state = Self {
            frames: Vec::with_capacity(MAX_FRAMES),
            stack: Vec::with_capacity(STACK_SIZE),
            globals: HashMap::new(),
            closures: Closures::default(),
            open_upvalues: Vec::with_capacity(STACK_SIZE),
        };
        state.define_native(strings, "clock", NativeFn(clock));
        state
    }
    fn push(&mut self, v: Value) {
        self.stack.push(v);
    }
    fn pop(&mut self) -> Value {
        self.stack.pop().expect("Empty stack")
    }

    fn peek(&self, n: usize) -> Value {
        let size = self.stack.len();
        self.stack[size - 1 - n]
    }

    fn define_native(&mut self, strings: &mut Strings, name: &str, native: NativeFn) {
        let name_id = strings.intern(name);
        self.globals.insert(name_id, Value::NativeFunction(native));
    }
}

#[derive(Default)]
pub struct Vm {
    strings: Strings,
    functions: Functions,
}

impl Vm {
    pub fn new_state(&mut self) -> ExecutionState {
        ExecutionState::new(&mut self.strings)
    }

    pub fn interpret(&mut self, code: &str, state: &mut ExecutionState) -> Result<(), LoxError> {
        let parser = Parser::new(code, &mut self.strings, &mut self.functions);
        let function = parser.compile()?;
        let closure = Closure::new(function);
        let closure_id = state.closures.store(closure);
        state.frames.push(CallFrame::new(function, closure_id));
        self.run(state)
    }

    fn binary_op<T>(
        &self,
        frame: &CallFrame,
        state: &mut ExecutionState,
        f: fn(f64, f64) -> T,
        r: fn(T) -> Value,
    ) -> Result<(), LoxError> {
        let operands = (state.pop(), state.pop());
        match operands {
            (Value::Number(b), Value::Number(a)) => {
                state.push(r(f(a, b)));
                Ok(())
            }
            _ => Err(self.runtime_error(&frame, "Operands must be numbers.")),
        }
    }

    fn run(&mut self, state: &mut ExecutionState) -> Result<(), LoxError> {
        let mut frame = state.frames.pop().unwrap();
        let mut chunk = {
            let closure = state.closures.lookup(frame.closure);
            &self.functions.lookup(closure.function).chunk
        };
        loop {
            let instruction = chunk.code[frame.ip];
            #[cfg(debug_assertions)]
            {
                for val in state.stack.iter() {
                    print!("[{}]", val);
                }
                println!("");
                #[cfg(debug_assertions)]
                chunk.disassemble_instruction(&instruction, frame.ip);
            }
            frame.ip += 1;

            match instruction {
                Instruction::Add => {
                    let (b, a) = (state.pop(), state.pop());
                    match (&a, &b) {
                        (Value::Number(a), Value::Number(b)) => {
                            state.push(Value::Number(a + b));
                        }
                        (Value::String(a), Value::String(b)) => {
                            let s_a = self.strings.lookup(*a);
                            let s_b = self.strings.lookup(*b);
                            let result = format!("{}{}", s_a, s_b);
                            let s = self.strings.intern_owned(result);
                            let value = Value::String(s);
                            state.push(value);
                        }
                        _ => {
                            state.push(a);
                            state.push(b);
                            return Err(self.runtime_error(
                                &frame,
                                "Operands must be numbers or string at the same time.",
                            ));
                        }
                    }
                }
                Instruction::CloseUpvalue => {
                    let stack_top = state.stack.len() - 1;
                    self.close_upvalues(state, stack_top);
                    state.pop();
                }
                Instruction::Closure(index) => {
                    let c = chunk.read_constant(index);
                    if let Value::Function(function_id) = c {
                        let function = self.functions.lookup(function_id);
                        let mut new_closure = Closure::new(function_id);

                        for upvalue in function.upvalues.iter() {
                            let obj_upvalue = if upvalue.is_local {
                                self.capture_value(state, frame.slot + upvalue.index)
                            } else {
                                let current_closure = state.closures.lookup(frame.closure);
                                current_closure.upvalues[upvalue.index].clone()
                            };
                            new_closure.upvalues.push(obj_upvalue);
                        }

                        let closure_id = state.closures.store(new_closure);
                        state.push(Value::Closure(closure_id));
                    }
                }
                Instruction::Call(arg_count) => {
                    frame = self.call_value(frame, state, arg_count)?;
                    chunk = &self.functions.lookup(frame.function).chunk;
                }
                Instruction::Constant(index) => {
                    let val = chunk.read_constant(index);
                    state.push(val);
                }
                Instruction::DefineGlobal(index) => {
                    let s = chunk.read_string(index);
                    let value = state.pop();
                    // no string, only the intern index
                    state.globals.insert(s, value);
                }
                Instruction::Divide => {
                    self.binary_op(&frame, state, |a, b| a / b, |n| Value::Number(n))?
                }
                Instruction::Equal => {
                    let a = state.pop();
                    let b = state.pop();
                    state.push(Value::Bool(a == b));
                }
                Instruction::False => state.push(Value::Bool(false)),
                Instruction::GetGlobal(index) => {
                    // just get the intern index
                    let s = chunk.read_string(index);
                    match state.globals.get(&s) {
                        Some(&value) => state.push(value),
                        None => {
                            let name = self.strings.lookup(s);
                            let msg = format!("Undefined variable '{}'.", name);
                            return Err(self.runtime_error(&frame, &msg));
                        }
                    }
                }
                Instruction::GetLocal(slot) => {
                    let i = slot + frame.slot;
                    let value = state.stack[i];
                    state.push(value);
                }
                Instruction::GetUpvalue(slot) => {
                    let value = {
                        let current_closure = state.closures.lookup(frame.closure);
                        let upvalue = current_closure.upvalues[slot].borrow();
                        if let Some(value) = upvalue.closed {
                            value
                        } else {
                            state.stack[upvalue.location]
                        }
                    };
                    state.push(value);
                }
                Instruction::Greater => {
                    self.binary_op(&frame, state, |a, b| a > b, |n| Value::Bool(n))?
                }
                Instruction::Jump(offset) => {
                    frame.ip += offset as usize;
                }
                Instruction::JumpIfFalse(offset) => {
                    if state.peek(0).is_falsy() {
                        frame.ip += offset as usize;
                    }
                }
                Instruction::Less => {
                    self.binary_op(&frame, state, |a, b| a < b, |n| Value::Bool(n))?
                }
                Instruction::Loop(offset) => {
                    frame.ip -= offset as usize + 1;
                }
                Instruction::Multiply => {
                    self.binary_op(&frame, state, |a, b| a * b, |n| Value::Number(n))?
                }
                Instruction::Negate => {
                    if let Value::Number(value) = state.peek(0) {
                        state.pop();
                        state.push(Value::Number(-value));
                    } else {
                        return Err(self.runtime_error(&frame, "Operand must be a number"));
                    }
                }
                Instruction::Nil => state.push(Value::Nil),
                Instruction::Not => {
                    let value = state.pop();
                    state.push(Value::Bool(value.is_falsy()));
                }
                Instruction::Pop => {
                    state.pop();
                }
                Instruction::Print => {
                    let value = state.pop();
                    if let Value::String(idx) = value {
                        println!("{}", self.strings.lookup(idx));
                    } else {
                        println!("{}", value);
                    }
                }


                Instruction::Return => {
                    let value = state.pop();
                    self.close_upvalues(state, frame.slot);
                    match state.frames.pop() {
                        Some(f) => {
                            state.stack.truncate(frame.slot);
                            state.push(value);
                            frame = f;
                            chunk = &self.functions.lookup(frame.function).chunk;
                        }
                        None => {
                            return Ok(());
                        }
                    }
                }
                Instruction::SetGlobal(index) => {
                    let name = chunk.read_string(index);
                    let value = state.peek(0);
                    // means not exist name, so this is an error
                    if let None = state.globals.insert(name, value) {
                        // delete
                        state.globals.remove(&name);
                        let s = self.strings.lookup(name);
                        let msg = format!("Undefined variable '{}'.", s);
                        return Err(self.runtime_error(&frame, &msg));
                    }
                }
                Instruction::SetLocal(slot) => {
                    let value = state.peek(0);
                    let i = slot + frame.slot;
                    state.stack[i] = value;
                }
                Instruction::SetUpvalue(slot) => {
                    let current_closure = state.closures.lookup(frame.closure);
                    let mut upvalue = current_closure.upvalues[slot].borrow_mut();
                    let value = state.peek(0);
                    if let None = upvalue.closed {
                        state.stack[upvalue.location] = value;
                    } else {
                        upvalue.closed = Some(value);
                    }
                }
                Instruction::Substract => {
                    self.binary_op(&frame, state, |a, b| a - b, |n| Value::Number(n))?
                }
                Instruction::True => state.push(Value::Bool(true)),
            }
        }
    }

    fn close_upvalues(&self, state: &mut ExecutionState, last: usize) {
        let mut i = 0;
        while i != state.open_upvalues.len() {
            if state.open_upvalues[i].borrow().location >= last {
                let upvalue = state.open_upvalues.remove(i);
                let location = upvalue.borrow().location;
                upvalue.borrow_mut().closed = Some(state.stack[location]);
            } else {
                i += 1;
            }
        }
    }

    fn capture_value(
        &self,
        state: &mut ExecutionState,
        location: usize,
    ) -> Rc<RefCell<ObjUpvalue>> {
        for upvalue in state.open_upvalues.iter() {
            if upvalue.borrow().location == location {
                return Rc::clone(upvalue);
            }
        }
        let upvalue = ObjUpvalue::new(location);
        let upvalue = Rc::new(RefCell::new(upvalue));
        state.open_upvalues.push(Rc::clone(&upvalue));
        upvalue
    }

    fn call_value(
        &self,
        frame: CallFrame,
        state: &mut ExecutionState,
        arg_count: usize,
    ) -> Result<CallFrame, LoxError> {
        let callee = state.peek(arg_count);
        match callee {
            Value::Closure(cid) => self.call(frame, state, cid, arg_count),
            Value::NativeFunction(native) => {
                let left = state.stack.len() - arg_count;
                let result = native.0(&state.stack[left..]);
                state.push(result);
                Ok(frame)
            }
            _ => Err(self.runtime_error(&frame, "Can only call function and classes.")),
        }
    }

    fn call(
        &self,
        frame: CallFrame,
        state: &mut ExecutionState,
        cid: ClosureID,
        arg_count: usize,
    ) -> Result<CallFrame, LoxError> {
        let closure = state.closures.lookup(cid);
        let f = self.functions.lookup(closure.function);
        if f.arity != arg_count {
            let msg = format!("Expected {} arguments but got {}.", f.arity, arg_count);
            Err(self.runtime_error(&frame, &msg))
        } else if state.frames.len() == MAX_FRAMES {
            Err(self.runtime_error(&frame, "Stack overflow."))
        } else {
            state.frames.push(frame);
            let mut frame = CallFrame::new(closure.function, cid);
            //println!("{} {}", state.stack.len(), arg_count);
            frame.slot = state.stack.len() - arg_count - 1;
            Ok(frame)
        }
    }

    fn runtime_error(&self, frame: &CallFrame, msg: &str) -> LoxError {
        eprintln!("{}", msg);
        let chunk = &self.functions.lookup(frame.function).chunk;
        let line = chunk.lines[frame.ip - 1];
        eprintln!("[line {}] in script", line);
        LoxError::RuntimeError
    }
}
